From 6cefa94088f497f5df611d7f067fa1f955e9a4f5 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Fri, 29 Jun 2018 10:10:24 +0800
Subject: [PATCH 4145/5242] MLK-18609-7 usb: cdns3: gadget: only unmap buffer
 for demand request

commit  53585b704e9e6c0f69e709b5cbfcc66b987ed656 from
https://source.codeaurora.org/external/imx/linux-imx.git

We only unmap the request which is demanded from the gadget driver.

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/gadget.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/cdns3/gadget.c b/drivers/usb/cdns3/gadget.c
index 0d866dc..57782a2 100644
--- a/drivers/usb/cdns3/gadget.c
+++ b/drivers/usb/cdns3/gadget.c
@@ -1870,9 +1870,6 @@ static int usb_ss_gadget_ep_dequeue(struct usb_ep *ep,
 	spin_lock_irqsave(&usb_ss->lock, flags);
 	dev_dbg(&usb_ss->dev, "DEQUEUE(%02X) %d\n",
 		ep->address, request->length);
-	usb_gadget_unmap_request_by_dev(usb_ss->sysdev, request,
-		ep->address & USB_DIR_IN);
-	request->status = -ECONNRESET;
 
 	select_ep(usb_ss, ep->desc->bEndpointAddress);
 	if (usb_ss->start_gadget)
@@ -1881,6 +1878,9 @@ static int usb_ss_gadget_ep_dequeue(struct usb_ep *ep,
 	list_for_each_entry_safe(req, req_temp,
 		&usb_ss_ep->request_list, list) {
 		if (request == req) {
+			request->status = -ECONNRESET;
+			usb_gadget_unmap_request_by_dev(usb_ss->sysdev, request,
+				ep->address & USB_DIR_IN);
 			list_del_init(&request->list);
 			if (request->complete) {
 				spin_unlock(&usb_ss->lock);
-- 
1.7.9.5

